Translate PostgreSQL custom aggregates

roji opened the pull request:
Here's some nice WIP based off of @smitpatel's great work in https://github.com/dotnet/efcore/pull/28092.

This implements the PG-specific aggregate operators ([FILTER](https://github.com/npgsql/efcore.pg/pull/2383/files#diff-93dee472cc58de12f790841bc6d992751b2af17d1fa3d7be5a3165bdbbb1aaa7R2465) and [ORDER BY](https://github.com/npgsql/efcore.pg/pull/2383/files#diff-93dee472cc58de12f790841bc6d992751b2af17d1fa3d7be5a3165bdbbb1aaa7R3158)), and provides a first implementation of [string_agg](https://github.com/npgsql/efcore.pg/pull/2383/files#diff-93dee472cc58de12f790841bc6d992751b2af17d1fa3d7be5a3165bdbbb1aaa7R3177) and [array_agg](https://github.com/npgsql/efcore.pg/pull/2383/files#diff-93dee472cc58de12f790841bc6d992751b2af17d1fa3d7be5a3165bdbbb1aaa7R3188). Really impressive we can do this!

/cc @dotnet/efteam

Closes #2384

smitpatel commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlQueryableAggregateMethodTranslator.cs:
```diff
@@ -0,0 +1,169 @@
+        MethodInfo method,
+        EnumerableExpression source,
+        IReadOnlyList<SqlExpression> arguments,
+        IDiagnosticsLogger<DbLoggerCategory.Query> logger)
```
Why do you need this here? 🤔

smitpatel replied:
Yes. I didn't notice the subtle difference that translation is calling into `AggregateFunction` method. Fine to override if you are generating a different translation. I thought it was same code as relational. In the absence of provider translating these methods, relational would provide it.

roji commented on src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlStatisticsAggregateMethodTranslator.cs:
```diff
@@ -39,18 +39,66 @@ public class NpgsqlStatisticsAggregateMethodTranslator : IAggregateMethodCallTra
+        {
+            // These methods accept two enumerable (column) arguments; this is represented in LINQ as a projection from the grouping
+            // to a tuple of the two columns. Since we generally translate tuples to PostgresRowValueExpression, we take it apart here.
+            if (source.Selector is not PostgresRowValueExpression rowValueExpression)
```
@smitpatel here's a translation of an aggregate function with multiple enumerable arguments. See test below for the C# side - as discussed the grouping needs to be projected to a tuple containing the two columns. It looks a bit weird/complex, but it prevents applying different predicate/orderings/distincts which would have been possible had the function accepted two enumerables instead.

What do you think, how does it look?

Funnily enough, since the PG provider translates ValueTuple to PostgresRowValueExpression, that's what we get here. That's also a bit weird, but seems OK.

smitpatel replied:
I do have some idea, let's discuss next time we are in office.

roji replied:
BTW @smitpatel at some point we can discuss this too

roji commented on test/EFCore.PG.FunctionalTests/Query/NorthwindFunctionsQueryNpgsqlTest.cs:
```diff
@@ -401,6 +402,165 @@ public virtual async Task GroupBy_JsonbAgg(bool async)
+                g => new
+                {
+                    City = g.Key,
+                    Companies = EF.Functions.JsonbObjectAgg<string, string, Dictionary<string, string>>(
```
And probably my last aggregate work for now :relieved: 

We now support the PG [jsonb_object_agg](https://www.postgresql.org/docs/current/functions-aggregate.html#FUNCTIONS-AGGREGATE-TABLE) function, using it to project out a `Dictionary<string,string>` out of two arbitrary columns in a table. Can also get a string JSON document (see test above).

(@ajcvickers this goes in the direction of projecting raw JSON data so we can send it directly over HTTP without entity dematerialization).

/cc @smitpatel @maumar @ajcvickers

roji commented:
Rebased on the merged aggregate infra support (#2393), cleaned up and split spatial aggregate out to #2384 (because it relies on upstream tests in https://github.com/dotnet/efcore/issues/28104#issuecomment-1142427678, which hasn't yet been merged).
